src/powershell/Program.cs

```diff
@@ -19,7 +20,236 @@ public sealed class ManagedPSEntry
+                { 
+                    continue;
+                }
+
+                // Check for "-Login" or some prefix thereof
+                if (IsParam(arg, "login", "LOGIN"))
+                {
+                    loginIndex = i;
+                    return true;
+                }
+
+                // After -File and -Command, all parameters are passed
+                // to the invoked file or command, so we can stop looking.
+                if (IsParam(arg, "file", "FILE")
+                    || IsParam(arg, "command", "COMMAND"))
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
+            {
+                return false;
+            }
+
+            // Check arg chars in order and allow prefixes
+            for (int i = 1; i < arg.Length; i++)
+            {
+                if (arg[i] != paramToCheck[i-1]
+                    && arg[i] != paramToCheckUpper[i-1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Create the exec call to /bin/{ba}sh -l -c 'pwsh "$@"' and run it.
+        /// </summary>
+        /// <param name="args">The argument vector passed to pwsh.</param>
+        /// <param name="loginArgIndex">The index of -Login in the argument vector.</param>
+        /// <returns>
+        /// The exit code of exec if it fails.
+        /// If exec succeeds, this process is overwritten so we never actually return.
+        /// </returns>
+        private static int ExecPwshLogin(string[] args, int loginArgIndex)
+        {
+            // We need the path to the current pwsh executable
+            string pwshPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName;
+
+            // Create input for /bin/sh that execs pwsh
+            int quotedPwshPathLength = GetQuotedPathLength(pwshPath);
+            string pwshInvocation = string.Create(
+                10 + quotedPwshPathLength, // exec '{pwshPath}' "$@"
+                (pwshPath, quotedPwshPathLength),
+                CreatePwshInvocation);
```

Comment:
I wonder that we do the optimization (and others) if we fork.

Reply:
Not sure I follow you here -- can you rephrase maybe?

Reply:
All the code run only once - no need to optimize something. I think we get more short and readable code if we will use classic string operations and string.Format.

Reply:
Right, the merits of various optimisations are dubious since we are just trying to exec as quickly as possible, which will then overwrite the whole process.

But I'm assuming that calls to things like malloc and the possibility of GC are still things we want to avoid. Similarly, if string formatting is handled by complex code in another assembly, I'd prefer to hardcode the small case here in an attempt to do our best to execute only the IL we know about here.